File: NoSQL330/discussion-module/src/main/java/by/rest/discussion/service/CommentService.java
package by.rest.discussion.service;

import by.rest.discussion.domain.Comment;
import by.rest.discussion.dto.CommentRequestTo;
import by.rest.discussion.dto.CommentResponseTo;
import by.rest.discussion.exception.ApiException;
import by.rest.discussion.mapper.CommentMapper;
import by.rest.discussion.repository.CommentRepository;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class CommentService {
    
    private final CommentRepository repository;
    private final CommentMapper mapper;
    
    public CommentService(CommentRepository repository, CommentMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }
    
    public CommentResponseTo create(CommentRequestTo request) {
        validate(request);
        
        Long storyId = request.getStoryId();
        Long nextId = generateNextId(storyId);
        
        Comment comment = new Comment(storyId, nextId, request.getContent());
        comment = repository.save(comment);
        
        return mapper.toResponse(comment);
    }
    
    // Временный упрощенный метод
    @Transactional(readOnly = true)
    public List<CommentResponseTo> getAllAsList() {
        try {
            return repository.findAll()
                    .stream()
                    .map(mapper::toResponse)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            e.printStackTrace();
            return List.of();
        }
    }
    
    // Временный упрощенный метод для storyId
    @Transactional(readOnly = true)
    public List<CommentResponseTo> getByStoryIdAsList(Long storyId) {
        try {
            return repository.findByStoryId(storyId)
                    .stream()
                    .map(mapper::toResponse)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            e.printStackTrace();
            return List.of();
        }
    }
    
    @Transactional(readOnly = true)
    public Page<CommentResponseTo> getAll(int page, int size) {
        try {
            Pageable pageable = PageRequest.of(page, size);
            Slice<Comment> commentsSlice = repository.findAll(pageable);
            
            List<CommentResponseTo> responses = commentsSlice.getContent()
                    .stream()
                    .map(mapper::toResponse)
                    .collect(Collectors.toList());
            
            long total = repository.count();
            return new PageImpl<>(responses, pageable, total);
        } catch (Exception e) {
            e.printStackTrace();
            return Page.empty();
        }
    }
    
    @Transactional(readOnly = true)
    public Page<CommentResponseTo> getByStoryId(Long storyId, int page, int size) {
        try {
            Pageable pageable = PageRequest.of(page, size);
            Slice<Comment> commentsSlice = repository.findByStoryId(storyId, pageable);
            
            List<CommentResponseTo> responses = commentsSlice.getContent()
                    .stream()
                    .map(mapper::toResponse)
                    .collect(Collectors.toList());
            
            long total = repository.countByStoryId(storyId);
            return new PageImpl<>(responses, pageable, total);
        } catch (Exception e) {
            e.printStackTrace();
            return Page.empty();
        }
    }
    
    @Transactional(readOnly = true)
    public CommentResponseTo getById(Long storyId, Long commentId) {
        Comment comment = repository.findByStoryIdAndId(storyId, commentId)
                .orElseThrow(() -> new ApiException(404, "40401", 
                    String.format("Comment not found with storyId=%d, id=%d", storyId, commentId)));
        
        return mapper.toResponse(comment);
    }
    
    public CommentResponseTo update(Long storyId, Long commentId, CommentRequestTo request) {
        validate(request);
        
        Comment comment = repository.findByStoryIdAndId(storyId, commentId)
                .orElseThrow(() -> new ApiException(404, "40401", 
                    String.format("Comment not found with storyId=%d, id=%d", storyId, commentId)));
        
        comment.setContent(request.getContent());
        comment = repository.save(comment);
        
        return mapper.toResponse(comment);
    }
    
    public void delete(Long storyId, Long commentId) {
        Comment.CommentKey key = new Comment.CommentKey(storyId, commentId);
        if (!repository.existsById(key)) {
            throw new ApiException(404, "40401", 
                String.format("Comment not found with storyId=%d, id=%d", storyId, commentId));
        }
        
        repository.deleteByStoryIdAndId(storyId, commentId);
    }
    
    private void validate(CommentRequestTo request) {
        if (request.getStoryId() == null || request.getStoryId() <= 0) {
            throw new ApiException(400, "40002", "Story ID must be positive");
        }
        
        if (request.getContent() == null || request.getContent().trim().isEmpty()) {
            throw new ApiException(400, "40002", "Content is required");
        }
        
        if (request.getContent().length() < 2 || request.getContent().length() > 2048) {
            throw new ApiException(400, "40002", "Content must be between 2 and 2048 characters");
        }
    }
    
    private Long generateNextId(Long storyId) {
        Long maxId = repository.findMaxIdByStoryId(storyId);
        return maxId != null ? maxId + 1 : 1L;
    }
}

File: 351002/Bubich/t310/app/services/comment_service.py
from typing import List
from app.models.comment import Comment
from app.dto.requests.comment_request import CommentRequestTo
from app.dto.responses.comment_response import CommentResponseTo
from app.repository.in_memory_repository import InMemoryRepository
from app.exceptions.custom_exceptions import NotFoundError, ValidationError


class CommentService:
    def __init__(self, repository: InMemoryRepository):
        self.repository = repository

    def create(self, request: CommentRequestTo) -> CommentResponseTo:
        self._validate_request(request)
        comment = Comment(
            story_id=request.story_id,
            content=request.content
        )
        saved_comment = self.repository.save(comment)
        return self._to_response(saved_comment)

    def get_by_id(self, id: int) -> CommentResponseTo:
        comment = self.repository.find_by_id(id)
        if not comment:
            raise NotFoundError(f"Comment with id {id} not found")
        return self._to_response(comment)

    def get_all(self) -> List[CommentResponseTo]:
        comments = self.repository.find_all()
        return [self._to_response(c) for c in comments]

    def update(self, id: int, request: CommentRequestTo) -> CommentResponseTo:
        self._validate_request(request)
        existing_comment = self.repository.find_by_id(id)
        if not existing_comment:
            raise NotFoundError(f"Comment with id {id} not found")

        existing_comment.story_id = request.story_id
        existing_comment.content = request.content
        updated_comment = self.repository.update(existing_comment)
        return self._to_response(updated_comment)

    def delete(self, id: int) -> None:
        if not self.repository.delete_by_id(id):
            raise NotFoundError(f"Comment with id {id} not found")

    def get_by_story_id(self, story_id: int) -> List[CommentResponseTo]:
        all_comments = self.repository.find_all()
        return [self._to_response(c) for c in all_comments if c.story_id == story_id]

    def _validate_request(self, request: CommentRequestTo):
        if not request.content or len(request.content) < 2 or len(request.content) > 2048:
            raise ValidationError("Content must be between 2 and 2048 characters")

    def _to_response(self, comment: Comment) -> CommentResponseTo:
        return CommentResponseTo(
            id=comment.id,
            story_id=comment.story_id,
            content=comment.content
        )